Fix errors on 1.7

darsnack opened the pull request:
`DataType` no longer has a `mutable` field. The new function `ismutabletype` can be used instead. Similarly, `abstract` is now `isabstracttype`.

`jl_new_datatype` now takes in a `fattrs` argument which I think should be `T.atomicfields` (maybe @vtjnash can correct me). The change was added in https://github.com/JuliaLang/julia/pull/37847

~~Only remaining error is that accessing the field `N` for `Vararg{T}` is now an undefined reference.~~

Fixes #107

darsnack commented:
Okay all the errors should have been resolved @DhairyaLGandhi

DhairyaLGandhi commented on src/extensions.jl:
```diff
@@ -86,7 +86,7 @@ tags[:array] = d ->
 
 structdata(x) = isprimitive(typeof(x)) ? reinterpret_(UInt8, [x]) :
-    Any[getfield(x, f) for f in fieldnames(typeof(x))]
+    Any[getfield(x, f) for f in fieldnames(typeof(x)) if isdefined(x, f)]
```
we can potentially run into issues here when trying to construct back the types, right?

darsnack replied:
If the field is uninitialized at this point, then there must be a constructor that excludes that field from its args right?

darsnack replied:
I think this should be fine because that argument would just not be there on re-construction. Which is fine cause how would it be uninitialized otherwise?

vtjnash replied:
Constructors don't in general map to the fields. The set of isdefined fields is frequently also not contiguous.

darsnack commented:
It seems we were only running into that issue for `Vararg{T}`, so I added a specific path for just that case and reverted the `isdefined` change. I also added some more tests for undefined references. There is one case that is currently broken (i.e. #43).

@vtjnash can you comment on what the `fattrs` argument to `jl_new_datatype` is supposed to represent? Seems like field attributes (e.g. atomic), but I wasn't sure what I should be passing in.

vtjnash commented:
It is unlikely that you should be trying to access internal functions via ccall. The allocator for `struct` objects is an intrinsic function: `Core._structtype(Main, :A, %3, %4, %5, false, 0)` (see `@Meta.lower struct A; end`), though the other functions called there are also mandatory, so calling `eval` with a struct definition is perhaps the most reliable version.

darsnack commented:
@CarloLucibello I don't have commit rights here, can you review?

The fixes here touch internals, but it is no more hacky than the rest of the package in that regard. A full overhaul to make the package more stable in the long term is needed (we already see failures on nightly), but not a task that I can complete in the near future. I would suggest this is merged so that BSON.jl is usable on 1.7.

darsnack commented:
We should tag a patch release too

CarloLucibello commented:
done
